### PR TITLE
Add back CUDA conflicts for GCC and Clang + Add CUDA 11.4.3 and 11.4.4

### PR DESCRIPTION
**Pull request by albestro**
In https://github.com/spack/spack/pull/28439 a couple of lines with still valid conflicts has been removed.

From https://github.com/spack/spack/pull/28439#issuecomment-1015448823 it seems that something changed from https://github.com/spack/spack/pull/25054#issuecomment-885543526. For instance, picking the line about CUDA 11.4.0 the two sources above says:

```
11.4.0-devel-ubuntu18.04/host_config.h:#error -- unsupported clang version! clang version must be less than 13 and greater than 3.2
11.4.0-devel-ubuntu18.04/host_config.h:#error -- unsupported GNU version! gcc versions later than 11 are not supported!
```

so "supported CLANG < 13 and GCC <= 11", while the other one says

```
11.4.0  #error -- unsupported clang version! clang version must be less than 12 and greater than 3.2
11.4.0  #error -- unsupported GNU version! gcc versions later than 10 are not supported!
```

so "supported CLANG < 12 and GCC <= 10".

Anyhow, from actual evidence I had (thanks @msimberg for pointing that out), at least for CUDA 11.2.0 and GCC 11 the conflicts is still there.

For this reason, after coordinating with @haampie, I start this PR by adding back the old conflicts for both GCC and CLANG, waiting for updates about the correct version of CUDA to use (is it 11.4.0 or do we have to use the older 11.3.1?).

Thanks for the support @haampie.

### CODE REVIEW

**Comment by albestro** [comment]
After a call with @haampie, this is the result of our investigation.

### TL;DR
Official CUDA docker image does not seem consistent with the actual CUDA toolkit installed in it (it reports 11.4.0, but probably the patch version is not correct). It is better to extract information from the cuda installation packages.

### Report
> From [#28439 (comment)](https://github.com/spack/spack/pull/28439#issuecomment-1015448823) it seems that something changed from [#25054 (comment)](https://github.com/spack/spack/pull/25054#issuecomment-885543526).

The information were extracted from the official nVidia docker image `nvidia/cuda:11.4.0-devel-ubuntu20.04`, which is tagged with CUDA version 11.4.0, but at two different times. Probably the docker image has been updated with the latest UPDATE of the 11.4 CUDA toolkit, but the version of the docker image does not report the correct patch version.


By inspecting the official CUDA toolkits (see note below on how it can be done), i.e. the ones available at the URLs used by spack for downloading the installer packages, these are the results:

```bash
cuda_11.4.0_470.42.01_linux.run.extract/builds/cuda_nvcc/targets/x86_64-linux/include/crt/host_config.h
#error -- unsupported GNU version! gcc versions later than 10 are not supported! The nvcc flag '-allow-unsupported-compiler' can be used to override this version check; however, using an unsupported host compiler may cause compilation failure or incorrect run time execution. Use at your own risk.
cuda_11.4.1_470.57.02_linux.run.extract/builds/cuda_nvcc/targets/x86_64-linux/include/crt/host_config.h
#error -- unsupported GNU version! gcc versions later than 11 are not supported! The nvcc flag '-allow-unsupported-compiler' can be used to override this version check; however, using an unsupported host compiler may cause compilation failure or incorrect run time execution. Use at your own risk.
cuda_11.4.2_470.57.02_linux.run.extract/builds/cuda_nvcc/targets/x86_64-linux/include/crt/host_config.h
#error -- unsupported GNU version! gcc versions later than 11 are not supported! The nvcc flag '-allow-unsupported-compiler' can be used to override this version check; however, using an unsupported host compiler may cause compilation failure or incorrect run time execution. Use at your own risk.
cuda_11.4.3_470.82.01_linux.run.extract/builds/cuda_nvcc/targets/x86_64-linux/include/crt/host_config.h
#error -- unsupported GNU version! gcc versions later than 11 are not supported! The nvcc flag '-allow-unsupported-compiler' can be used to override this version check; however, using an unsupported host compiler may cause compilation failure or incorrect run time execution. Use at your own risk.
cuda_11.4.4_470.82.01_linux.run.extract/builds/cuda_nvcc/targets/x86_64-linux/include/crt/host_config.h
#error -- unsupported GNU version! gcc versions later than 11 are not supported! The nvcc flag '-allow-unsupported-compiler' can be used to override this version check; however, using an unsupported host compiler may cause compilation failure or incorrect run time execution. Use at your own risk.
```

```bash
cuda_11.4.0_470.42.01_linux.run.extract/builds/cuda_nvcc/targets/x86_64-linux/include/crt/host_config.h
#error -- unsupported clang version! clang version must be less than 12 and greater than 3.2 . The nvcc flag '-allow-unsupported-compiler' can be used to override this version check; however, using an unsupported host compiler may cause compilation failure or incorrect run time execution. Use at your own risk.
cuda_11.4.1_470.57.02_linux.run.extract/builds/cuda_nvcc/targets/x86_64-linux/include/crt/host_config.h
#error -- unsupported clang version! clang version must be less than 13 and greater than 3.2 . The nvcc flag '-allow-unsupported-compiler' can be used to override this version check; however, using an unsupported host compiler may cause compilation failure or incorrect run time execution. Use at your own risk.
cuda_11.4.2_470.57.02_linux.run.extract/builds/cuda_nvcc/targets/x86_64-linux/include/crt/host_config.h
#error -- unsupported clang version! clang version must be less than 13 and greater than 3.2 . The nvcc flag '-allow-unsupported-compiler' can be used to override this version check; however, using an unsupported host compiler may cause compilation failure or incorrect run time execution. Use at your own risk.
cuda_11.4.3_470.82.01_linux.run.extract/builds/cuda_nvcc/targets/x86_64-linux/include/crt/host_config.h
#error -- unsupported clang version! clang version must be less than 13 and greater than 3.2 . The nvcc flag '-allow-unsupported-compiler' can be used to override this version check; however, using an unsupported host compiler may cause compilation failure or incorrect run time execution. Use at your own risk.
cuda_11.4.4_470.82.01_linux.run.extract/builds/cuda_nvcc/targets/x86_64-linux/include/crt/host_config.h
#error -- unsupported clang version! clang version must be less than 13 and greater than 3.2 . The nvcc flag '-allow-unsupported-compiler' can be used to override this version check; however, using an unsupported host compiler may cause compilation failure or incorrect run time execution. Use at your own risk.
```

As it can be seen, 11.4.0 still has the "previous" conflict as reported in the older comment, while starting from 11.4.1 there is the updated compatibility as reported in the newer comment.

> Anyhow, from actual evidence I had (thanks @msimberg for pointing that out), at least for CUDA 11.2.0 and GCC 11 the conflicts is still there.
> 
> For this reason, after coordinating with @haampie, I start this PR by adding back the old conflicts for both GCC and CLANG, waiting for updates about the correct version of CUDA to use (is it 11.4.0 or do we have to use the older 11.3.1?).

So, the conclusion would be that the old conflict has to be restored and updated to just include 11.4.0, since from 11.4.1 the compiler compatibility has been upgraded.

Moreover, for future reference, we want also to report that from [CUDA release notes](https://docs.nvidia.com/cuda/cuda-toolkit-release-notes/index.html#cuda-major-component-versions__table-cuda-toolkit-driver-versions) and the [CUDA toolkit archive](https://developer.nvidia.com/cuda-toolkit-archive), it seems that the "UPDATE" matches with the "patch" component in the semantic versioning.

For completeness, I will add in this PR also missing CUDA versions entries for 11.4.3 and 11.4.4.

### Note: extracting "host_config.h" from CUDA Toolkit installers

CUDA package installers are [makeself](https://makeself.io/) scripts. These are, more or less, a bundle containing the installer script and the archive to be installed. In order to avoid extracting them just to look at a single file, it is possible to use the `--tar` option to control what the extraction process should do. In particular, I used something like

```bash
bash cuda_11.4.4_470.82.01_linux.run --tar "xf" "-C $outdir" "--wildcards *nvcc*host_config.h"
```

that will result in just the `host_config.h` file extracted in the `$outdir` folder (`-C $outdir` is optional, but it is helpful when dealing with multiple toolkits at the same time).

**Comment by albestro** [comment]
I was quickly looking at the other PRs...what about https://github.com/spack/spack/pull/25054#pullrequestreview-746548769 @ax3l?

Let's wait to collect all the information before fixing the conflict. 😉

**Comment by albestro** [comment]
> I was quickly looking at the other PRs...what about [#25054 (review)](https://github.com/spack/spack/pull/25054#pullrequestreview-746548769) @ax3l?
> 
> Let's wait to collect all the information before fixing the conflict. 😉

On Monday, I will try to check better. Just reporting here something that may be useful for the discussion.

@ax3l Following the thread of your comment, I read that the problem with GCC10 and CUDA11.4.0 was about something like `chrono` in stdlib...and it rings me a bell.

Indeed, working with `umpire+cuda %gcc@10.3.0 ^ cuda@11.2.0, I addressed a problem that may be the same...but my conclusion was that it was a problem of GCC (see https://github.com/LLNL/Umpire/issues/653 and https://github.com/spack/spack/pull/26028)

My conclusion is the result of a search that lead me to this [GCC bug](https://gcc.gnu.org/bugzilla/show_bug.cgi?id=100102#c25) for which there should be a patch for GCC10.3.0, and from the comment it is

> Fixed for GCC 10.4, 11.2 and 12.

In order to confirm this, I will try to build `umpire+cuda%gcc@10.4.0` with both `^ cuda@11.2.0` and `^ cuda@11.4.0`. If it works, it would support my hypothesis and we can probably go for just representing the conflicts as per information collected from cuda toolkits `host_config.h` files.

In addition to that, we may also decide to test the GCC patch for 10.3.0, as it was already suggested in https://github.com/spack/spack/pull/26028#issuecomment-931481845. However, about this last point, I would need help for better defining the conflict, since it will be a conflict for "plain" 10.3.0, but it should work for the one with the specified patch.

File: lib/spack/spack/build_systems/cuda.py [hunk 1]
```diff
@@ -106,7 +106,9 @@ def cuda_flags(arch_list):
         # This implies that the last one in the list has to be updated at
         # each release of a new cuda minor version.
         conflicts('%gcc@10:', when='+cuda ^cuda@:11.0')
+        conflicts('%gcc@11:', when='+cuda ^cuda@:11.4')
```
Comment (haampie) [inline lib/spack/spack/build_systems/cuda.py]:
```suggestion
        conflicts('%gcc@11:', when='+cuda ^cuda@:11.4.0')
```

File: lib/spack/spack/build_systems/cuda.py [hunk 2]
```diff
@@ -106,7 +106,9 @@ def cuda_flags(arch_list):
...
         conflicts('%gcc@12:', when='+cuda ^cuda@:11.6')
+        conflicts('%clang@12:', when='+cuda ^cuda@:11.4')
```
Comment (haampie) [inline lib/spack/spack/build_systems/cuda.py]:
```suggestion
        conflicts('%clang@12:', when='+cuda ^cuda@:11.4.0')
```

**Review by haampie — APPROVED** [review]
Can you copy your `.run`-file trick to the issue / github gist so that we can get back to it next time? should probably automate it...

**Comment by ax3l** [comment]
Sorry I did not see this earlier, too many deadlines & proposals early in the year...
